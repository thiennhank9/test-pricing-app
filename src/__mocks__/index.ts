import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import axiosMockAdapter from "axios-mock-adapter";
import api from "api";

const mockApi = new axiosMockAdapter(api);
const middleware = [thunk];
const mockStore = configureMockStore(middleware);

export { mockApi, mockStore };
