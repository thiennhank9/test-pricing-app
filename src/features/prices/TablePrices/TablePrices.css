.table-prices {
  width: 100%;
}

.table-container {
  position: relative;
}

.table-spinner {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  opacity: 0.5;
  background-color: #c5c6d0;
}

.row-button-see-more {
  display: flex;
  justify-content: flex-end;
  margin-top: 15px;
}

.error-text {
  color: red;
}
