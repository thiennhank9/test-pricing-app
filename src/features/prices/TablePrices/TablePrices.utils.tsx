import { PriceData } from "../pricesTypes";
import { isNil } from "utils";

export interface CellInfo {
  cell: PriceData;
  rowIndex?: number;
  colIndex?: number;
}

export const topHeaders = [
  "Quantity",
  "Day 1",
  "Day 2",
  "Day 3",
  "Day 4",
  "Day 5",
];

export const defaultCellInfo: CellInfo = {
  cell: {},
  rowIndex: undefined,
  colIndex: undefined,
};

const defaultDisplayData = new Array(5).fill(new Array(5).fill({}));

export const getDisplayPrices = (
  prices: Array<any>,
  showAll: boolean
): Array<any> => {
  if (!prices.length) {
    return defaultDisplayData;
  }
  return showAll ? prices : prices.slice(0, 5);
};

export const getLeftHeaders = (prices: Array<Array<PriceData>>) =>
  prices.map((prices: Array<PriceData>) => prices[0].quantity);

export const defaultPriceData: PriceData = {
  price: undefined,
  businessDay: undefined,
  quantity: undefined,
};

export function getCellClassName(
  cellInfo: CellInfo,
  hover: CellInfo,
  selected: CellInfo
): string {
  if (isNil(hover.rowIndex) && isNil(selected.rowIndex)) {
    return "";
  }
  const isSelected =
    selected.colIndex === cellInfo.colIndex &&
    selected.rowIndex === cellInfo.rowIndex;
  if (isSelected) {
    return "high-light";
  }
  const isHover =
    hover.colIndex === cellInfo.colIndex &&
    hover.rowIndex === cellInfo.rowIndex;
  if (isHover) {
    return "high-light";
  }
  const isSameColumnOrRowWithHoverCell =
    hover.colIndex === cellInfo.colIndex ||
    hover.rowIndex === cellInfo.rowIndex;
  if (isSameColumnOrRowWithHoverCell) {
    return "weak";
  }
  return "";
}
