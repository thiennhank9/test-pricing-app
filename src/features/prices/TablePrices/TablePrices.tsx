/* eslint-disable react-hooks/exhaustive-deps */
import React, { useState, useEffect, useMemo, useCallback } from "react";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "app/store";
import { selectPrice, getPrices } from "../pricesSlice";
import Button from "components/Button";
import Table from "components/Table";
import { PriceData } from "../pricesTypes";
import {
  topHeaders,
  getDisplayPrices,
  getLeftHeaders,
  getCellClassName,
  CellInfo,
  defaultCellInfo,
  defaultPriceData,
} from "./TablePrices.utils";
import "./TablePrices.css";
import Spinner from "components/Spinner";
import { formatCurrency } from "utils";

function TablePrices() {
  const dispatch = useDispatch();
  const { prices, paperSize, error, loading } = useSelector(
    (state: RootState) => state.prices
  );
  const [selected, setSelected] = useState(defaultCellInfo);
  const [hover, setHover] = useState(defaultCellInfo);
  const [showAll, setShowAll] = useState(false);
  const disabledHoverAndSelect = useMemo(
    () => prices.length === 0,
    [prices.length]
  );
  const displayPrices = useMemo(
    () => getDisplayPrices(prices, showAll),
    [prices, showAll]
  );
  const leftHeaders = useMemo(
    () => getLeftHeaders(displayPrices),
    [displayPrices]
  );

  // Functions
  const clearHover = useCallback(() => setHover(defaultCellInfo), []);
  const clearSelected = useCallback(() => setSelected(defaultCellInfo), []);
  const unSelectCell = useCallback(() => {
    clearSelected();
    dispatch(selectPrice(defaultPriceData));
  }, []);
  const onClickCell = (cellInfo: CellInfo) => {
    const isSelected =
      cellInfo.rowIndex === selected.rowIndex &&
      cellInfo.colIndex === selected.colIndex;
    if (isSelected) {
      unSelectCell();
      return;
    }
    dispatch(selectPrice(cellInfo.cell));
    setSelected(cellInfo);
  };
  const renderCell = (cell: PriceData, rowIndex: number, colIndex: number) => {
    const cellInfo = { cell, rowIndex, colIndex };
    const cellKey = `${rowIndex}-${colIndex}`;
    const onClick = () => onClickCell(cellInfo);
    const onMouseOver = () => setHover(cellInfo);
    const cellClassName = disabledHoverAndSelect
      ? ""
      : getCellClassName(cellInfo, hover, selected);
    return (
      <td
        key={cellKey}
        onClick={onClick}
        onMouseOver={onMouseOver}
        className={cellClassName}
      >
        {formatCurrency(cell.price)}
      </td>
    );
  };

  // Effects
  useEffect(() => {
    clearHover();
    clearSelected();
    dispatch(getPrices(paperSize));
  }, [paperSize]);

  return (
    <div className="table-prices">
      <h3>Price table</h3>
      <p>Paper size: {paperSize}</p>
      <div className="table-container">
        <Table
          topHeaders={topHeaders}
          leftHeaders={leftHeaders}
          data={displayPrices}
          renderCell={renderCell}
          onMouseOutTableBody={clearHover}
        />
        {loading && (
          <div className="table-spinner">
            <Spinner />
          </div>
        )}
      </div>
      <div className="row-button-see-more">
        <Button onClick={() => setShowAll(!showAll)}>
          {showAll ? "See less" : "See more"}
        </Button>
      </div>
      <div className="error-text">{error}</div>
    </div>
  );
}

export default TablePrices;
