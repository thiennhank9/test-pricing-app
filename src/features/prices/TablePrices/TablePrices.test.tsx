import Spinner from "components/Spinner";
import React from "react";
import { Provider } from "react-redux";
import TestRenderer from "react-test-renderer";
import { mockStore } from "__mocks__";
import TablePrices from "./TablePrices";

describe("TablePrices", () => {
  test("render table prices with correct paper size", () => {
    const store = mockStore({
      prices: {
        paperSize: "A4",
        prices: [],
      },
    });
    const testRenderer = TestRenderer.create(
      <Provider store={store}>
        <TablePrices />
      </Provider>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
    const testInstance = testRenderer.root;
    const text = testInstance.findByType("p");
    expect(text.props.children.includes("A4")).toBe(true);
  });
  test("table prices has loading", () => {
    const store = mockStore({
      prices: {
        paperSize: "A4",
        prices: [],
        loading: true
      },
    });
    const testRenderer = TestRenderer.create(
      <Provider store={store}>
        <TablePrices />
      </Provider>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
    const testInstance = testRenderer.root;
    const spinner = testInstance.findByType(Spinner);
    expect(spinner).toBeDefined();
  });
  test("table prices has error", () => {
    const store = mockStore({
      prices: {
        paperSize: "A4",
        prices: [],
        error: "error"
      },
    });
    const testRenderer = TestRenderer.create(
      <Provider store={store}>
        <TablePrices />
      </Provider>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
    const testInstance = testRenderer.root;
    const errorText = testInstance.findByProps({className: "error-text"});
    expect(errorText.props.children).toBe("error");
  });
});
