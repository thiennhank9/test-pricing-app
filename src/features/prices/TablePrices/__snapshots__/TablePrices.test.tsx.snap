// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TablePrices render table prices with correct paper size 1`] = `
<div
  className="table-prices"
>
  <h3>
    Price table
  </h3>
  <p>
    Paper size: 
    A4
  </p>
  <div
    className="table-container"
  >
    <table>
      <thead>
        <tr>
          <th>
            Quantity
          </th>
          <th>
            Day 1
          </th>
          <th>
            Day 2
          </th>
          <th>
            Day 3
          </th>
          <th>
            Day 4
          </th>
          <th>
            Day 5
          </th>
        </tr>
      </thead>
      <tbody
        onMouseOut={[Function]}
      >
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="row-button-see-more"
  >
    <button
      onClick={[Function]}
    >
      See more
    </button>
  </div>
  <div
    className="error-text"
  />
</div>
`;

exports[`TablePrices table prices has error 1`] = `
<div
  className="table-prices"
>
  <h3>
    Price table
  </h3>
  <p>
    Paper size: 
    A4
  </p>
  <div
    className="table-container"
  >
    <table>
      <thead>
        <tr>
          <th>
            Quantity
          </th>
          <th>
            Day 1
          </th>
          <th>
            Day 2
          </th>
          <th>
            Day 3
          </th>
          <th>
            Day 4
          </th>
          <th>
            Day 5
          </th>
        </tr>
      </thead>
      <tbody
        onMouseOut={[Function]}
      >
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="row-button-see-more"
  >
    <button
      onClick={[Function]}
    >
      See more
    </button>
  </div>
  <div
    className="error-text"
  >
    error
  </div>
</div>
`;

exports[`TablePrices table prices has loading 1`] = `
<div
  className="table-prices"
>
  <h3>
    Price table
  </h3>
  <p>
    Paper size: 
    A4
  </p>
  <div
    className="table-container"
  >
    <table>
      <thead>
        <tr>
          <th>
            Quantity
          </th>
          <th>
            Day 1
          </th>
          <th>
            Day 2
          </th>
          <th>
            Day 3
          </th>
          <th>
            Day 4
          </th>
          <th>
            Day 5
          </th>
        </tr>
      </thead>
      <tbody
        onMouseOut={[Function]}
      >
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
        <tr>
          <th />
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
          <td
            className=""
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            
          </td>
        </tr>
      </tbody>
    </table>
    <div
      className="table-spinner"
    >
      <div
        className="spinner"
      />
    </div>
  </div>
  <div
    className="row-button-see-more"
  >
    <button
      onClick={[Function]}
    >
      See more
    </button>
  </div>
  <div
    className="error-text"
  />
</div>
`;
